components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImpl.java

```diff
@@ -852,6 +838,91 @@ public ServiceProvider getServiceProviderByClientId(String clientId, String clie
         return serviceProvider;
     }
 
+    public ImporterResponse importSPApplication(String content, String fileName, String tenantDomain, String username,
+                                                boolean isUpdate) throws IdentityApplicationManagementException {
+
+        ImporterResponse importerResponse = new ImporterResponse();
+        ServiceProvider serviceProvider = unmarshalSP(content, fileName, tenantDomain);
+        if (serviceProvider == null) {
+            throw new IdentityApplicationManagementException(String.format("Empty Service Provider file %s upload by" +
+                    " tenant: %s", fileName, tenantDomain));
+        }
+
+
+        Collection<ApplicationMgtListener> listeners =
+                ApplicationMgtListenerServiceComponent.getApplicationMgtListeners();
+
+        ServiceProvider savedSP = null;
+        try {
+            if (isUpdate) {
+                savedSP = getApplicationExcludingFileBasedSPs(serviceProvider.getApplicationName(), tenantDomain);
+                if (savedSP == null) {
+                    if (log.isDebugEnabled()) {
+                        log.debug(String.format("Service provider %s@%s is not found, so creating new Service provider",
+                                serviceProvider.getApplicationName(), tenantDomain));
+                    }
+                    isUpdate = false;
+                }
+            }
+
+            if (!isUpdate) {
+                ServiceProvider basicApplication = new ServiceProvider();
+                basicApplication.setApplicationName(serviceProvider.getApplicationName());
+                basicApplication.setDescription(serviceProvider.getDescription());
+                savedSP = addApplication(basicApplication, tenantDomain, username);
+            }
+            serviceProvider.setApplicationID(savedSP.getApplicationID());
+            serviceProvider.setOwner(getUser(tenantDomain, username));
+
+            for (ApplicationMgtListener listener : listeners) {
+                if (listener.isEnable()) {
+                    listener.doPreUpdateApplication(serviceProvider, tenantDomain, username);
+                }
+            }
+            isImportSP.set(true);
```

Comment:
As discussed offline, move this to become the immediate before to updateApplication call.